webodf/lib/ops/OperationTransformMatrix.js

```diff
@@ -274,6 +274,262 @@ ops.OperationTransformMatrix = function OperationTransformMatrix() {
+                }
+            }
+            // 2, because 1 for pos inside annotation comment, 1 for new pos before annotated range
+            insertTextSpec.position += 2;
+        }
+
+        return {
+            opSpecsA:  [addAnnotationSpec],
+            opSpecsB:  [insertTextSpec]
+        };
+    }
+
+    /**
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpec
+     * @param {!ops.OpMergeParagraph.Spec} mergeParagraphSpec
+     * @return {?{opSpecsA:!Array.<!Object>, opSpecsB:!Array.<!Object>}}
+     */
+    function transformAddAnnotationMergeParagraph(addAnnotationSpec, mergeParagraphSpec) {
+        if (mergeParagraphSpec.sourceStartPosition <= addAnnotationSpec.position) {
+            addAnnotationSpec.position -= 1;
+        } else {
+            if (addAnnotationSpec.length !== undefined) {
+                if (mergeParagraphSpec.sourceStartPosition <= addAnnotationSpec.position + addAnnotationSpec.length) {
+                    addAnnotationSpec.length -= 1;
+                }
+            }
+
+            // 2, because 1 for pos inside annotation comment, 1 for new pos before annotated range
+            mergeParagraphSpec.sourceStartPosition += 2;
+
+            if (addAnnotationSpec.position < mergeParagraphSpec.destinationStartPosition) {
+                mergeParagraphSpec.destinationStartPosition += 2;
+            }
+        }
+
+        return {
+            opSpecsA:  [addAnnotationSpec],
+            opSpecsB:  [mergeParagraphSpec]
+        };
+    }
+
+    /**
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpec
+     * @param {!ops.OpMoveCursor.Spec} moveCursorSpec
+     * @return {?{opSpecsA:!Array.<!Object>, opSpecsB:!Array.<!Object>}}
+     */
+    function transformAddAnnotationMoveCursor(addAnnotationSpec, moveCursorSpec) {
+        var isMoveCursorSpecRangeInverted = invertMoveCursorSpecRangeOnNegativeLength(moveCursorSpec);
+
+        // adapt movecursor spec to inserted positions
+        if (addAnnotationSpec.position < moveCursorSpec.position) {
+            // 2, because 1 for pos inside annotation comment, 1 for new pos before annotated range
+            moveCursorSpec.position += 2;
+        } else if (addAnnotationSpec.position < moveCursorSpec.position + moveCursorSpec.length) {
+            // 2, because 1 for pos inside annotation comment, 1 for new pos before annotated range
+            moveCursorSpec.length += 2;
+        }
+
+        if (isMoveCursorSpecRangeInverted) {
+            invertMoveCursorSpecRange(moveCursorSpec);
+        }
+
+        return {
+            opSpecsA:  [addAnnotationSpec],
+            opSpecsB:  [moveCursorSpec]
+        };
+    }
+
+    /**
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpec
+     * @param {!ops.OpRemoveAnnotation.Spec} removeAnnotationSpec
+     * @return {?{opSpecsA:!Array.<!Object>, opSpecsB:!Array.<!Object>}}
+     */
+    function transformAddAnnotationRemoveAnnotation(addAnnotationSpec, removeAnnotationSpec) {
```

Comment:
I tried to distill an action proposal out of this, but failed. Because it was in a homeopathic dose? :)